src/current/v24.1/changefeed-messages.md

```diff
@@ -196,20 +196,22 @@ CREATE CHANGEFEED FOR TABLE rides INTO 'external://kafka' WITH diff, envelope=wr
 
 Changefeeds provide the following guarantees for message delivery to changefeed sinks:
 
-- [Per-key ordering](#per-key-ordering)
-- [At-least-once delivery](#at-least-once-delivery)
+- [Per-key ordering](#per-key-ordering) for the first emission of an event's message.
+- [At-least-once delivery](#at-least-once-delivery) per event message.
 
 {{site.data.alerts.callout_info}}
 Changefeeds do not support total message ordering or transactional ordering of messages.
 {{site.data.alerts.end}}
 
 ### Per-key ordering
 
-Changefeeds provide a _per-key ordering guarantee_ for messages emitted to the sink. Once the changefeed has emitted a row with a timestamp, the changefeed will not emit any previously unseen versions of that row with a lower timestamp. Therefore, you will never receive a **new** change for that row at an earlier timestamp.
+Changefeeds provide a _per-key ordering guarantee_ for the **first emission** of a message to the sink. Once the changefeed has emitted a row with a timestamp, the changefeed will not emit any previously unseen versions of that row with a lower timestamp. Therefore, you will never receive a new change for that row at an earlier timestamp.
 
 For example, a changefeed can emit updates to rows `A` at timestamp `T1`, `B` at `T2`, and `C` at `T3` in any order.
 
-When there are updates to rows `A` at `T1`, `B` at `T2`, and `A` at `T3`, the changefeed will always emit `A` at `T3` **after** `A` at `T1`. However, `A` at `T3` could precede or follow `B` at `T2`. This is because there is no timestamp ordering between keys.
+When there are updates to rows `A` at `T1`, `B` at `T2`, and `A` at `T3`, the changefeed will always emit `A` at `T3` (for the first time) **after** emitting `A` at `T1` (for the first time). However, `A` at `T3` could precede or follow `B` at `T2`, because there is no timestamp ordering between keys.
```

Comment:
if the thing catching these messages at the other end is conceptually like a hash table (* HUGE ASSUMPTION ofc) it doesn't seem like this would be a problem e.g.

```
changes['A']['T1'] = 'foo.diff'
changes['A']['T2'] = 'bar.diff'
```
so if the changefeed later emits the same updates to rows `A` and `T2` again, I don't really have to care bcs it's just saying 

```
changes['A']['T2'] = 'bar.diff'
```

which since it's just a dupe message with no changes to the actual diffed content would be a no-op, it seems

anyway don't mind me this is probably wrong, and I'm sure none of this works like a little hash table on my little computer, just thinking out loud about it a little

Reply:
I get your point! And I believe next steps are to provide some more practical guidance around re-emitted messages. 

I think this is what was not clear — whether all dupes would be emitted, and whether they would be in order. Depending on how the changefeed has batched some changes, I think there could potentially be unexpected results if all dupes in the correct order were expected.